src/sql/workbench/api/node/extHostModelView.ts

```diff
@@ -34,27 +35,50 @@ class ModelBuilderImpl implements sqlops.ModelBuilder {
 
 	card(): sqlops.ComponentBuilder<sqlops.CardComponent> {
 		let id = this.getNextComponentId();
-		return new ComponentBuilderImpl(new CardWrapper(this._proxy, this._handle, id));
+		return this.withEventHandler(new CardWrapper(this._proxy, this._handle, id), id);
+	}
+
+	inputBox(): sqlops.ComponentBuilder<sqlops.InputBoxComponent> {
+		let id = this.getNextComponentId();
+		return this.withEventHandler(new InputBoxWrapper(this._proxy, this._handle, id), id);
 	}
 
 	dashboardWidget(widgetId: string): sqlops.ComponentBuilder<sqlops.WidgetComponent> {
 		let id = this.getNextComponentId();
-		return new ComponentBuilderImpl<sqlops.WidgetComponent>(new ComponentWrapper(this._proxy, this._handle, ModelComponentTypes.DashboardWidget, id));
+		return this.withEventHandler<sqlops.WidgetComponent>(new ComponentWrapper(this._proxy, this._handle, ModelComponentTypes.DashboardWidget, id), id);
 	}
 
 	dashboardWebview(webviewId: string): sqlops.ComponentBuilder<sqlops.WebviewComponent> {
 		let id = this.getNextComponentId();
-		return new ComponentBuilderImpl(new ComponentWrapper(this._proxy, this._handle, ModelComponentTypes.DashboardWebview, id));
+		return this.withEventHandler(new ComponentWrapper(this._proxy, this._handle, ModelComponentTypes.DashboardWebview, id), id);
+	}
+
+	withEventHandler<T extends sqlops.Component>(component: ComponentWrapper, id: string): sqlops.ComponentBuilder<T> {
+		let componentBuilder: ComponentBuilderImpl<T> = new ComponentBuilderImpl<T>(component);
+		this._eventHandlers.set(id, componentBuilder);
+		return componentBuilder;
+	}
+
+	handleEvent(id: string, eventArgs: IComponentEventArgs): void {
+		let eventHandler = this._eventHandlers.get(id);
+		if (eventHandler) {
+			eventHandler.handleEvent(eventArgs);
+		}
 	}
 
 	private getNextComponentId(): string {
 		return `component${this._handle}_${this.nextComponentId++}`;
 	}
 }
 
-class ComponentBuilderImpl<T extends sqlops.Component> implements sqlops.ComponentBuilder<T> {
+interface IWithEventHandler {
+	handleEvent(eventArgs: IComponentEventArgs): void;
+}
+
+class ComponentBuilderImpl<T extends sqlops.Component> implements sqlops.ComponentBuilder<T>, IWithEventHandler {
 
 	constructor(protected _component: ComponentWrapper) {
+		_component.registerEvent();
```

Comment:
Do all components need to register for event handling then?

Reply:
I think  all components want to have this by default. I make it optional so the component to register or not 

---
In reply to: [183179896](https://github.com/Microsoft/sqlopsstudio/pull/1190#discussion_r183179896) [](ancestors = 183179896)